File: benchmarks-QuTiP/multiplication_dense_ket.py
import qutip as qt
import benchmarkutils

name = "multiplication_dense_ket"

samples = 2
evals = 5
cutoffs = range(100, 1001, 100)


def setup(N):
    op1 = qt.rand_dm(N, N) * 0.2j
    psi = qt.rand_ket(N)
    return op1, psi


def f(op1, psi):
    return op1 * psi


print("Benchmarking:", name)
print("Cutoff: ", end="", flush=True)
results = []
for N in cutoffs:
    print(N, "", end="", flush=True)
    op1, psi = setup(N)
    t = benchmarkutils.run_benchmark(f, op1, psi,
                                     samples=samples, evals=evals)
    results.append({"N": N, "t": t})
print()
benchmarkutils.save(name, results)

File: benchmarks-QuantumOptics.jl/multiplication_dense_ket.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

using Random; Random.seed!(0)

name = "multiplication_dense_ket"

samples = 2
evals = 5
cutoffs = [50:50:500;]

function setup(N)
    b = GenericBasis(N)
    op1 = randoperator(b)
    psi = randstate(b)
    result = copy(psi)
    op1, psi, result
end

function f(op1, psi, result)
    operators.gemv!(ComplexF64(1., 0.), op1, psi, ComplexF64(0., 0.), result)
end

println("Benchmarking: ", name)
print("Cutoff: ")
results = []
for N in cutoffs
    print(N, " ")
    op1, psi, result = setup(N)
    t = @belapsed f($op1, $psi, $result) samples=samples evals=evals
    push!(results, Dict("N"=>N, "t"=>t))
end
println()
benchmarkutils.save(name, results)
